### PR TITLE
Refactor Common SearchParameter Functionality

### PR DESCRIPTION
**Pull request by nathandoef**
- Moved validation logic performed inside `JpaResourceDaoSearchParameter` into a new class called `SearchParameterDaoValidator`
- Moved logic for extracting combo SearchParameters that was inside `SearchParamWithInlineReferencesExtractor` into `BaseSearchParamExtractor` where other types of SearchParameters are extracted

### CODE REVIEW

File: hapi-fhir-jpaserver-model/src/main/java/ca/uhn/fhir/jpa/model/entity/IResourceIndexComboSearchParameter.java [hunk 1]
```diff
@@ -0,0 +1,16 @@
+package ca.uhn.fhir.jpa.model.entity;
+
+import org.hl7.fhir.instance.model.api.IIdType;
+
+public interface IResourceIndexComboSearchParameter {
```
Comment (tadgh) [inline hapi-fhir-jpaserver-model/src/main/java/ca/uhn/fhir/jpa/model/entity/IResourceIndexComboSearchParameter.java:5]: issue: add Javadocs on the class as to why this exists. Methods are self evident though.

File: hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/util/ISearchParamRegistry.java [hunk 2]
```diff
@@ -73,6 +76,14 @@ default List<RuntimeSearchParam> getActiveComboSearchParams(String theResourceNa
 		return Collections.emptyList();
 	}
 
+	default List<RuntimeSearchParam> getActiveComboSearchParams(String theResourceName, ComboSearchParamType theParamType) {
+		return Collections.emptyList();
+	}
+
+	default Optional<RuntimeSearchParam> getActiveComboSearchParamById(String theResourceName, IIdType theId) {
+		return Optional.empty();
+	}
```
Comment (tadgh) [inline hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/util/ISearchParamRegistry.java:85]: issue: i see james has a TODO above to remove default methods. I think these are only present to help backwards compat with smile until we can implement over there. I recommend leaving yourself a similar todo, and once all is merged on both sides, clearing thesse default implentatsions in a subsequent MR.

File: hapi-fhir-storage/src/main/java/ca/uhn/fhir/jpa/dao/validation/SearchParameterDaoValidator.java [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+			return;
+		}
+
+		if (ElementUtil.isEmpty(searchParameter.getBase()) && (searchParameter.getType() == null || !Enumerations.SearchParamType.COMPOSITE.name().equals(searchParameter.getType().name()))) {
+			throw new UnprocessableEntityException(Msg.code(1113) + "SearchParameter.base is missing");
+		}
+
+		boolean isUnique = hasAnyExtensionUniqueSetTo(searchParameter, true);
+
+		if (searchParameter.getType() != null && searchParameter.getType().name().equals(Enumerations.SearchParamType.COMPOSITE.name()) && isBlank(searchParameter.getExpression())) {
```
Comment (tadgh) [inline hapi-fhir-storage/src/main/java/ca/uhn/fhir/jpa/dao/validation/SearchParameterDaoValidator.java:70]: suggestion(non-blocking): take these triple-argument conditionals and break them into more human-readable methods, e.g. `isCompositeWithoutExpression()`

**Review by tadgh — APPROVED** [review]
Approved pending some minor requests.. Awesome work! This does need a changelog though, for HAPI-FHIR API users.

File: hapi-fhir-jpaserver-model/src/main/java/ca/uhn/fhir/jpa/model/entity/IResourceIndexComboSearchParameter.java (hunk 1 above)
Comment (nathandoef) [reply hapi-fhir-jpaserver-model/src/main/java/ca/uhn/fhir/jpa/model/entity/IResourceIndexComboSearchParameter.java:10]: Good call, fixed

File: hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/util/ISearchParamRegistry.java (hunk 2 above)
Comment (nathandoef) [reply hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/util/ISearchParamRegistry.java:88]: Added TODOs

File: hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/dao/JpaResourceDaoSearchParameterTest.java [hunk 4]
```diff
@@ -39,10 +43,13 @@ public void before() {
 
 		myDao = new JpaResourceDaoSearchParameter<>();
 		myDao.setContext(myCtx);
+		DaoConfig config = new DaoConfig();
```
[the remark this answers is not in the document]
Comment (nathandoef) [reply hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/dao/JpaResourceDaoSearchParameterTest.java]: Fixed

File: hapi-fhir-storage/src/main/java/ca/uhn/fhir/jpa/dao/validation/SearchParameterDaoValidator.java (hunk 3 above)
Comment (nathandoef) [reply hapi-fhir-storage/src/main/java/ca/uhn/fhir/jpa/dao/validation/SearchParameterDaoValidator.java]: Fixed

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/index/SearchParamWithInlineReferencesExtractor.java [hunk 5]
```diff
@@ -146,124 +133,12 @@ public void populateFromResource(RequestPartitionId theRequestPartitionId, Resou
...
-						}
-					}
-				}
-			}
-		}
-
-		return ResourceIndexedSearchParams.extractCompositeStringUniquesValueChains(theResourceType, partsChoices);
+	private void extractComboParameters(ResourceTable theEntity, ResourceIndexedSearchParams theParams) {
+		mySearchParamExtractorService.extractSearchParamComboUnique(theEntity, theParams);
+		mySearchParamExtractorService.extractSearchParamComboNonUnique(theEntity, theParams);
```
[the remark this answers is not in the document]
Comment (nathandoef) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/index/SearchParamWithInlineReferencesExtractor.java:141]: Might be better to keep this name to match the style of `ISearchParamExtractor`. Thoughts?

 https://github.com/hapifhir/hapi-fhir/pull/4321/files#diff-dd271df620b0bce8aa0ae1aa2a982810d1903007f9d7041e4a542614d21a2d33

[the remark this answers is not in the document]
Comment (tadgh) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/index/SearchParamWithInlineReferencesExtractor.java:141]: ah fair enough 👍🏻

**Pull request merged**